rftp/file_response.go

```diff
@@ -45,11 +49,27 @@ func newFileResponse(index uint16) *FileResponse {
 		pwriter:       w,
 		buffer:        newChunkQueue(index),
 		resendEntries: make(map[uint64]struct{}),
+		hasher:        md5.New(),
 	}
 }
 
 func (f *FileResponse) Read(p []byte) (n int, err error) {
-	return f.preader.Read(p)
+	n, err = f.preader.Read(p)
+	if err != nil {
+		return
```

Comment:
This `return` comes too early. `err` might be EOF and the documentation of [io.Reader](https://golang.org/pkg/io/#Reader) says you have to handle n>0 bytes before considering the error. Thus, we have to add n bytes to the hasher before returning the error, which may be EOF or some other io error.